File: r/tests/testthat/test-aggregate.R
## Fixtures: two states, four jurisdictions, so every count is checkable by hand.

panel_fixture <- function() {
  tibble::tibble(
    year = rep(2024L, 4),
    fips_code = c("0100100000", "0100300000", "2300100000", "23"),
    state_abbr = c("AL", "AL", "ME", "ME"),
    mail_rejected = c(10, 20, 5, NA),
    prov_rejected = c(1, NA, NA, NA)
  )
}

status_fixture <- function() {
  tibble::tibble(
    year = rep(2024L, 4),
    fips_code = c("0100100000", "0100300000", "2300100000", "23"),
    state_abbr = c("AL", "AL", "ME", "ME"),
    mail_rejected = c("reported", "reported", "reported", "not_available"),
    # AL: one reported, one genuinely missing. ME: both inapplicable.
    prov_rejected = c("reported", "not_available", "does_not_apply", "valid_skip")
  )
}

test_that("values are summed per state and the shape is long", {
  out <- eavs_aggregate(panel_fixture())
  expect_setequal(unique(out$concept), c("mail_rejected", "prov_rejected"))
  expect_equal(nrow(out), 2 * 2) # 2 states x 2 concepts
  expect_equal(out$value[out$concept == "mail_rejected" & out$state_abbr == "AL"], 30)
  expect_equal(out$value[out$concept == "mail_rejected" & out$state_abbr == "ME"], 5)
})

test_that("a state total equals a plain sum of its jurisdictions", {
  panel <- panel_fixture()
  out <- eavs_aggregate(panel)
  expect_equal(
    sum(out$value[out$concept == "mail_rejected"]),
    sum(panel$mail_rejected, na.rm = TRUE)
  )
})

test_that("every input row is accounted for in n_total", {
  panel <- panel_fixture()
  out <- eavs_aggregate(panel)
  expect_equal(sum(out$n_total[out$concept == "mail_rejected"]), nrow(panel))
})

test_that("without status every absence counts as a gap", {
  out <- eavs_aggregate(panel_fixture())
  me <- out[out$concept == "mail_rejected" & out$state_abbr == "ME", ]
  expect_false(me$coverage_exact)
  expect_equal(me$n_reported, 1L)
  expect_equal(me$n_missing, 1L)
  expect_equal(me$n_not_applicable, 0L)
  expect_equal(me$coverage, 0.5)
})

test_that("with status, does_not_apply is not counted as a gap", {
  out <- eavs_aggregate(panel_fixture(), status = status_fixture())
  me <- out[out$concept == "prov_rejected" & out$state_abbr == "ME", ]
  expect_true(me$coverage_exact)
  expect_equal(me$n_reported, 0L)
  expect_equal(me$n_missing, 0L)       # does_not_apply + valid_skip are not gaps
  expect_equal(me$n_not_applicable, 2L)
  expect_true(is.na(me$coverage))      # nothing could have been answered
})

test_that("blanks are counted apart and stay in the coverage denominator", {
  status <- status_fixture()
  status$mail_rejected <- c("reported", "blank", "reported", "not_available")
  out <- eavs_aggregate(panel_fixture(), status = status)
  al <- out[out$concept == "mail_rejected" & out$state_abbr == "AL", ]
  expect_equal(al$n_reported, 1L)
  expect_equal(al$n_missing, 0L)
  expect_equal(al$n_blank, 1L)
  # A blank counts as a gap in coverage, so the number reads as a lower bound.
  expect_equal(al$coverage, 0.5)
})

test_that("entity_type separates states, territories, and DC", {
  panel <- panel_fixture()
  panel$state_abbr <- c("AL", "PR", "DC", "ME")
  out <- eavs_aggregate(panel)
  types <- unique(out[, c("state_abbr", "entity_type")])
  expect_equal(types$entity_type[types$state_abbr == "AL"], "state")
  expect_equal(types$entity_type[types$state_abbr == "PR"], "territory")
  expect_equal(types$entity_type[types$state_abbr == "DC"], "district")
})

test_that("not_applicable is excluded from the coverage denominator", {
  out <- eavs_aggregate(panel_fixture(), status = status_fixture())
  al <- out[out$concept == "prov_rejected" & out$state_abbr == "AL", ]
  # One reported, one not_available: 1/2, not 1/2-of-something-larger.
  expect_equal(al$n_reported, 1L)
  expect_equal(al$n_missing, 1L)
  expect_equal(al$coverage, 0.5)
})

test_that("status changes coverage but never the summed value", {
  panel <- panel_fixture()
  without <- eavs_aggregate(panel)
  with <- eavs_aggregate(panel, status = status_fixture())
  expect_equal(without$value, with$value)
  expect_false(isTRUE(all.equal(without$n_missing, with$n_missing)))
})

test_that("a mismatched status frame is an error", {
  expect_error(
    eavs_aggregate(panel_fixture(), status = status_fixture()[1:2, ]),
    "line up row-for-row"
  )
})

test_that("concepts= restricts the rollup and rejects unknown names", {
  out <- eavs_aggregate(panel_fixture(), concepts = "mail_rejected")
  expect_equal(unique(out$concept), "mail_rejected")
  expect_error(eavs_aggregate(panel_fixture(), concepts = "nope"), "Not numeric concept")
})

test_that("a panel with no numeric concepts is an error", {
  bad <- tibble::tibble(year = 2024L, state_abbr = "AL", fips_code = "0100100000")
  expect_error(eavs_aggregate(bad), "No numeric concept columns")
})

test_that("a missing year column is an error", {
  bad <- panel_fixture()
  bad$year <- NULL
  expect_error(eavs_aggregate(bad), "needs a .*year.* column")
})

test_that("county rollups drop codes that embed no county, and say so", {
  jur <- tibble::tibble(
    year = rep(2024L, 4),
    fips_code = c("0100100000", "0100300000", "2300100000", "23"),
    county_fips = c("01001", "01003", "23001", NA) # Maine's statewide row has none
  )
  expect_message(
    out <- eavs_aggregate(panel_fixture(), by = "county", jurisdictions = jur),
    "Dropping 1 row"
  )
  expect_true("county_fips" %in% names(out))
  expect_equal(sum(out$n_total[out$concept == "mail_rejected"]), 3)
})

test_that("multiple years stay separate", {
  panel <- rbind(panel_fixture(), transform(panel_fixture(), year = 2022L))
  out <- eavs_aggregate(panel)
  expect_setequal(unique(out$year), c(2022L, 2024L))
  expect_equal(nrow(out), 2 * 2 * 2) # 2 years x 2 states x 2 concepts
})

test_that("coverage_reg weights coverage by reg_eligible_total", {
  panel <- panel_fixture()
  panel$reg_eligible_total <- c(1000, 3000, 500, 1500)
  out <- eavs_aggregate(panel)
  me <- out[out$concept == "mail_rejected" & out$state_abbr == "ME", ]
  expect_equal(me$coverage, 0.5)
  expect_equal(me$coverage_reg, 0.25) # 500 reported of 2000 answerable
  al <- out[out$concept == "mail_rejected" & out$state_abbr == "AL", ]
  expect_equal(al$coverage_reg, 1)
})

test_that("coverage_reg is NA without the weight, and NA weights drop from both sides", {
  out <- eavs_aggregate(panel_fixture())
  expect_true(all(is.na(out$coverage_reg)))
  panel <- panel_fixture()
  panel$reg_eligible_total <- c(1000, 3000, 500, NA)
  out <- eavs_aggregate(panel)
  me <- out[out$concept == "mail_rejected" & out$state_abbr == "ME", ]
  # The unreported Maine row has no usable weight, so it leaves the weighted
  # ratio entirely while the unweighted coverage still counts it as a gap.
  expect_equal(me$coverage, 0.5)
  expect_equal(me$coverage_reg, 1)
})

test_that("an NA reason counts as not collected, and the buckets sum to n_total", {
  status <- status_fixture()
  status$mail_rejected <- c("reported", "not_available", NA, NA)
  out <- eavs_aggregate(panel_fixture(), status = status)
  me <- out[out$concept == "mail_rejected" & out$state_abbr == "ME", ]
  expect_equal(me$n_not_collected, 2L)
  expect_equal(me$n_reported + me$n_missing + me$n_blank + me$n_not_applicable, 0L)
  expect_true(is.na(me$coverage))      # nothing was asked, so nothing to cover
  al <- out[out$concept == "mail_rejected" & out$state_abbr == "AL", ]
  expect_equal(al$n_not_collected, 0L)
  expect_equal(al$coverage, 0.5)
  sums <- out$n_reported + out$n_missing + out$n_blank +
    out$n_not_applicable + out$n_not_collected
  expect_equal(sums, out$n_total)
})

test_that("a status frame in the wrong year order is an error, not wrong coverage", {
  panel <- rbind(panel_fixture(), transform(panel_fixture(), year = 2022L))
  status <- rbind(status_fixture(), transform(status_fixture(), year = 2022L))
  expect_silent(out <- eavs_aggregate(panel, status = status))
  reversed <- status[c(5:8, 1:4), ]
  expect_equal(nrow(reversed), nrow(panel))   # the row count check cannot see it
  expect_error(eavs_aggregate(panel, status = reversed), "not aligned")
})

File: py/tests/test_aggregate.py
"""State and county rollups, and reason-aware coverage.

Mirrors r/tests/testthat/test-aggregate.R. Fixtures small enough to check every
count by hand. Network-free.
"""

from __future__ import annotations

import pandas as pd
import pytest

import tidyeavs

FIPS = ["0100100000", "0100300000", "2300100000", "23"]


def panel() -> pd.DataFrame:
    return pd.DataFrame(
        {
            "year": [2024] * 4,
            "fips_code": FIPS,
            "state_abbr": ["AL", "AL", "ME", "ME"],
            "mail_rejected": [10.0, 20.0, 5.0, None],
            "prov_rejected": [1.0, None, None, None],
        }
    )


def status() -> pd.DataFrame:
    return pd.DataFrame(
        {
            "year": [2024] * 4,
            "fips_code": FIPS,
            "state_abbr": ["AL", "AL", "ME", "ME"],
            "mail_rejected": ["reported", "reported", "reported", "not_available"],
            # AL: one reported, one genuine gap. ME: both inapplicable.
            "prov_rejected": ["reported", "not_available", "does_not_apply", "valid_skip"],
        }
    )


def cell(out, concept, state, column):
    row = out[(out["concept"] == concept) & (out["state_abbr"] == state)]
    return row[column].item()


def test_values_summed_per_state_and_output_is_long():
    out = tidyeavs.aggregate(panel())
    assert set(out["concept"]) == {"mail_rejected", "prov_rejected"}
    assert len(out) == 2 * 2
    assert cell(out, "mail_rejected", "AL", "value") == 30
    assert cell(out, "mail_rejected", "ME", "value") == 5


def test_state_total_equals_a_plain_sum():
    p = panel()
    out = tidyeavs.aggregate(p)
    assert out.loc[out["concept"] == "mail_rejected", "value"].sum() == p[
        "mail_rejected"
    ].sum()


def test_every_input_row_is_accounted_for():
    p = panel()
    out = tidyeavs.aggregate(p)
    assert out.loc[out["concept"] == "mail_rejected", "n_total"].sum() == len(p)


def test_without_status_every_absence_is_a_gap():
    out = tidyeavs.aggregate(panel())
    assert cell(out, "mail_rejected", "ME", "coverage_exact") is False or not cell(
        out, "mail_rejected", "ME", "coverage_exact"
    )
    assert cell(out, "mail_rejected", "ME", "n_reported") == 1
    assert cell(out, "mail_rejected", "ME", "n_missing") == 1
    assert cell(out, "mail_rejected", "ME", "n_not_applicable") == 0
    assert cell(out, "mail_rejected", "ME", "coverage") == 0.5


def test_does_not_apply_is_not_a_gap():
    out = tidyeavs.aggregate(panel(), status=status())
    assert bool(cell(out, "prov_rejected", "ME", "coverage_exact"))
    assert cell(out, "prov_rejected", "ME", "n_reported") == 0
    assert cell(out, "prov_rejected", "ME", "n_missing") == 0
    assert cell(out, "prov_rejected", "ME", "n_not_applicable") == 2
    assert pd.isna(cell(out, "prov_rejected", "ME", "coverage"))


def test_blanks_counted_apart_and_kept_in_coverage_denominator():
    st = status()
    st["mail_rejected"] = ["reported", "blank", "reported", "not_available"]
    out = tidyeavs.aggregate(panel(), status=st)
    assert cell(out, "mail_rejected", "AL", "n_reported") == 1
    assert cell(out, "mail_rejected", "AL", "n_missing") == 0
    assert cell(out, "mail_rejected", "AL", "n_blank") == 1
    # A blank counts as a gap in coverage, so the number reads as a lower bound.
    assert cell(out, "mail_rejected", "AL", "coverage") == 0.5


def test_entity_type_separates_states_territories_and_dc():
    p = panel()
    p["state_abbr"] = ["AL", "PR", "DC", "ME"]
    out = tidyeavs.aggregate(p)
    types = dict(zip(out["state_abbr"], out["entity_type"]))
    assert types["AL"] == "state"
    assert types["PR"] == "territory"
    assert types["DC"] == "district"


def test_not_applicable_excluded_from_coverage_denominator():
    out = tidyeavs.aggregate(panel(), status=status())
    assert cell(out, "prov_rejected", "AL", "n_reported") == 1
    assert cell(out, "prov_rejected", "AL", "n_missing") == 1
    assert cell(out, "prov_rejected", "AL", "coverage") == 0.5


def test_status_changes_coverage_but_never_the_value():
    without = tidyeavs.aggregate(panel())
    with_status = tidyeavs.aggregate(panel(), status=status())
    key = ["year", "state_abbr", "concept"]
    a = without.sort_values(key).reset_index(drop=True)
    b = with_status.sort_values(key).reset_index(drop=True)
    assert a["value"].equals(b["value"])
    assert not a["n_missing"].equals(b["n_missing"])


def test_mismatched_status_is_an_error():
    with pytest.raises(ValueError, match="line up row-for-row"):
        tidyeavs.aggregate(panel(), status=status().iloc[:2])


def test_concepts_filter_and_unknown_name():
    out = tidyeavs.aggregate(panel(), concepts=["mail_rejected"])
    assert set(out["concept"]) == {"mail_rejected"}
    with pytest.raises(ValueError, match="Not numeric concept"):
        tidyeavs.aggregate(panel(), concepts=["nope"])


def test_no_numeric_concepts_is_an_error():
    bad = pd.DataFrame({"year": [2024], "state_abbr": ["AL"], "fips_code": ["0100100000"]})
    with pytest.raises(ValueError, match="No numeric concept columns"):
        tidyeavs.aggregate(bad)


def test_missing_year_is_an_error():
    with pytest.raises(ValueError, match="needs a 'year' column"):
        tidyeavs.aggregate(panel().drop(columns="year"))


def test_bad_by_is_an_error():
    with pytest.raises(ValueError, match="by must be"):
        tidyeavs.aggregate(panel(), by="precinct")


def test_county_rollup_drops_codes_with_no_county():
    jur = pd.DataFrame(
        {
            "year": pd.array([2024] * 4, dtype="Int64"),
            "fips_code": pd.array(FIPS, dtype="string"),
            # Maine's statewide row embeds no county.
            "county_fips": pd.array(["01001", "01003", "23001", None], dtype="string"),
        }
    )
    with pytest.warns(UserWarning, match="Dropping 1 row"):
        out = tidyeavs.aggregate(panel(), by="county", jurisdictions=jur)
    assert "county_fips" in out.columns
    assert out.loc[out["concept"] == "mail_rejected", "n_total"].sum() == 3


def test_multiple_years_stay_separate():
    p = pd.concat([panel(), panel().assign(year=2022)], ignore_index=True)
    out = tidyeavs.aggregate(p)
    assert set(out["year"]) == {2022, 2024}
    assert len(out) == 2 * 2 * 2


def test_coverage_reg_weights_coverage_by_registration():
    p = panel()
    p["reg_eligible_total"] = [1000.0, 3000.0, 500.0, 1500.0]
    out = tidyeavs.aggregate(p)
    assert cell(out, "mail_rejected", "ME", "coverage") == 0.5
    assert cell(out, "mail_rejected", "ME", "coverage_reg") == 0.25
    assert cell(out, "mail_rejected", "AL", "coverage_reg") == 1.0


def test_coverage_reg_na_without_weight_and_na_weights_drop():
    out = tidyeavs.aggregate(panel())
    assert out["coverage_reg"].isna().all()
    p = panel()
    p["reg_eligible_total"] = [1000.0, 3000.0, 500.0, None]
    out = tidyeavs.aggregate(p)
    # The unreported Maine row has no usable weight, so it leaves the weighted
    # ratio entirely while the unweighted coverage still counts it as a gap.
    assert cell(out, "mail_rejected", "ME", "coverage") == 0.5
    assert cell(out, "mail_rejected", "ME", "coverage_reg") == 1.0


def test_na_reason_counts_as_not_collected_and_buckets_sum():
    st = status()
    st["mail_rejected"] = ["reported", "not_available", None, None]
    out = tidyeavs.aggregate(panel(), status=st)
    assert cell(out, "mail_rejected", "ME", "n_not_collected") == 2
    assert cell(out, "mail_rejected", "ME", "n_reported") == 0
    assert cell(out, "mail_rejected", "ME", "n_missing") == 0
    assert pd.isna(cell(out, "mail_rejected", "ME", "coverage"))
    assert cell(out, "mail_rejected", "AL", "n_not_collected") == 0
    assert cell(out, "mail_rejected", "AL", "coverage") == 0.5
    sums = (
        out["n_reported"]
        + out["n_missing"]
        + out["n_blank"]
        + out["n_not_applicable"]
        + out["n_not_collected"]
    )
    assert (sums == out["n_total"]).all()


def test_status_in_the_wrong_year_order_is_an_error():
    p = pd.concat([panel(), panel().assign(year=2022)], ignore_index=True)
    st = pd.concat([status(), status().assign(year=2022)], ignore_index=True)
    tidyeavs.aggregate(p, status=st)  # aligned: fine
    reversed_status = pd.concat([st.iloc[4:], st.iloc[:4]], ignore_index=True)
    assert len(reversed_status) == len(p)  # the row count check cannot see it
    with pytest.raises(ValueError, match="not aligned"):
        tidyeavs.aggregate(p, status=reversed_status)


# known_anomaly: a total can rest on an anomalous convention while every
# coverage column reads as complete, which is exactly Iowa's 2018 polling places
# (all 99 counties reporting).


def agg_anomalies() -> pd.DataFrame:
    return pd.DataFrame(
        {
            "year": [2024],
            "state_abbr": ["AL"],
            "concept": ["mail_rejected"],
            "note": ["Test fixture."],
            "source": ["Test fixture."],
        }
    )


def test_known_anomaly_marks_totals_resting_on_an_anomalous_state_year():
    out = tidyeavs.aggregate(panel(), anomalies=agg_anomalies())
    hit = (out["state_abbr"] == "AL") & (out["concept"] == "mail_rejected")
    assert out.loc[hit, "known_anomaly"].all()
    assert not out.loc[~hit, "known_anomaly"].any()
    # Coverage is unaffected: the column reports, it does not adjust.
    assert out.loc[hit, "coverage"].notna().all()


def test_known_anomaly_is_false_where_the_group_reported_nothing():
    frame = panel()
    frame.loc[frame["state_abbr"] == "AL", "mail_rejected"] = None
    out = tidyeavs.aggregate(frame, anomalies=agg_anomalies())
    hit = (out["state_abbr"] == "AL") & (out["concept"] == "mail_rejected")
    assert (out.loc[hit, "n_reported"] == 0).all()
    assert not out.loc[hit, "known_anomaly"].any()


def test_anomalies_none_skips_the_lookup():
    out = tidyeavs.aggregate(panel(), anomalies=None)
    assert not out["known_anomaly"].any()
